executor/index_merge_reader.go

```diff
@@ -0,0 +1,724 @@
+// 2. indexMergeProcessWorker, which is used to do the `Union` operation.
+// 3. indexMergeTableScanWorker, which is used to get the table tuples with the given handles.
+//
+// The execution flow is really like IndexLookUpReader. However, it uses multiple index scans
+// or table scans to get the handles:
+// 1. use the partialTableWorkers and partialIndexWorkers to fetch the handles (a batch per time)
+//    and send them to the indexMergeProcessWorker.
+// 2. indexMergeProcessWorker do the `Union` operation for a batch of handles it have got.
+//    For every handle in the batch:
+//	  1. check whether it has been accessed.
+//    2. if not, record it and send it to the indexMergeTableScanWorker.
+//    3. if accessed, just ignore it.
+type IndexMergeReaderExecutor struct {
+	baseExecutor
+
+	table      table.Table
+	indexes    []*model.IndexInfo
+	keepOrders []bool
+	descs      []bool
+	ranges     [][]*ranger.Range
+	dagPBs     []*tipb.DAGRequest
+	startTS    uint64
+	// handleIdx is the index of handle, which is only used for case of keeping order.
+	handleIdx    int
+	tableRequest *tipb.DAGRequest
+	// columns are only required by union scan.
+	columns           []*model.ColumnInfo
+	partialStreamings []bool
+	tableStreaming    bool
+	*dataReaderBuilder
+	// All fields above are immutable.
+
+	partialWorkerWg sync.WaitGroup
+	tblWorkerWg     sync.WaitGroup
+	processWokerWg  sync.WaitGroup
+	finished        chan struct{}
+
+	kvRanges []kv.KeyRange
+
+	resultCh   chan *lookupTableTask
+	resultCurr *lookupTableTask
+	feedbacks  []*statistics.QueryFeedback
+
+	// memTracker is used to track the memory usage of this executor.
+	memTracker *memory.Tracker
+
+	// checkIndexValue is used to check the consistency of the index data.
+	*checkIndexValue
+
+	corColInIdxSide bool
+	partialPlans    [][]plannercore.PhysicalPlan
+	corColInTblSide bool
+	tblPlans        []plannercore.PhysicalPlan
+	corColInAccess  bool
+	idxCols         [][]*expression.Column
+	colLens         [][]int
+}
+
+// Open implements the Executor Open interface
+func (e *IndexMergeReaderExecutor) Open(ctx context.Context) error {
+	kvRangess := make([][]kv.KeyRange, 0, len(e.partialPlans))
+	for i, plan := range e.partialPlans {
+		_, ok := plan[0].(*plannercore.PhysicalIndexScan)
+		if !ok {
+			kvRangess = append(kvRangess, nil)
+			continue
+		}
+		kvRanges, err := distsql.IndexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, getPhysicalTableID(e.table), e.indexes[i].ID, e.ranges[i], e.feedbacks[i])
+		if err != nil {
+			return err
+		}
+		kvRangess = append(kvRangess, kvRanges)
+	}
+
+	e.finished = make(chan struct{})
+	e.resultCh = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	workCh := make(chan *lookupTableTask, 1)
+	fetchCh := make(chan *lookupTableTask, len(kvRangess))
+
+	e.startIndexMergeProcessWorker(ctx, workCh, fetchCh, len(kvRangess))
+	e.partialWorkerWg.Add(len(kvRangess))
```

Comment:
It's better to put this statement into the loop below and increase it by one each time, otherwise, if there is something wrong before starting all goroutines, it may cause deadlock at `partialWorkerWg.Wait(0` when closing.